### ISSUE TITLE
Project or editor crashes randomly from xcb XInitThreads

### ISSUE THREAD

**Issue by Cykyrios** [godotengine/godot#75308]
### Godot version

v4.1.dev.custom_build [0291fcd7b]

### System information

Linux Manjaro, kernel 6.1.19, X11

### Issue description

The editor or the running project sometimes crashes with the following error:
```
[xcb] Unknown sequence number while processing queue
[xcb] You called XInitThreads, this is not your fault
[xcb] Aborting, sorry about that.
godot.linuxbsd.editor.x86_64: xcb_io.c:278: poll_for_event: Assertion `!xcb_xlib_threads_sequence_lost' failed.
```
Crashes are more common while a project is running, but the editor also crashed because of this a couple of times over the past week or so.

I am not using any thread-related functions in my project, physics/rendering are not threaded, the project I'm working on as this happens is a simple GUI-based game.

### Steps to reproduce

This seems to happen fairly randomly.

### Minimal reproduction project

N/A

**Comment by Calinou** [comment]
I haven't been able to reproduce this so far on Fedora 37 KDE (GeForce RTX 4090 with NVIDIA 525.89.02).

What graphics card model, driver version and desktop environment are you using?

**Comment by Cykyrios** [comment]
Oh right, I forgot about GPU-related info. I have an AMD 7900 XT, running on the open-source amdgpu drivers with Mesa 22.3.5 (amdgpu version is "kernel").
The desktop is Plasma 5.26.5.

**Comment by geowarin** [comment]
Saw this happening (only once) on a totally different config:
RTX 2080Ti
archlinux
i3wm

**Comment by cg9999** [comment]
I have this too, again. Seems very reminiscent of #69352
Happens quite frequently here.
The message varies a bit, last one I got is:
```
[xcb] Unknown request in queue while dequeuing
[xcb] You called XInitThreads, this is not your fault
[xcb] Aborting, sorry about that.
godot: xcb_io.c:175: dequeue_pending_request: Assertion `!xcb_xlib_unknown_req_in_deq' failed.
```

Godot: v4.0.1.stable.arch_linux
libx11: 1.8.4-1
arch linux 64 bit, kernel 6.2.8-zen1
On a laptop with Intel HD Graphics 620

**Comment by DrRevert** [comment]
Manjaro kernel version: 6.1.22-1
As for GPUs: AMD RX 6800 XT but also Intel UHD Graphics 770 (CPU i7-12700)
I have my screens connected to integrated graphics as I'm doing some GPU passthrough, this setup used to cause issues during the beta whenever I opened a new window or a submenu.
Happened like 4 times mostly randomly when the editor was idling, one time when I tried to press the button to Git commit in editor, but that's probably a coincidence.

**Comment by Eraph** [comment]
Seeing the same thing on Manjaro here, I have integrated Intel graphics. Common factor seems to be Arch based distros?

**Comment by vypxl** [comment]
Having the same issue, Manjaro with Hyprland / Wayland here. Also Godot 4.0.1 stable, libx11 v1.8.4-1, intel integrated graphics.

**Comment by ilmagico** [comment]
Can confirm on Manjaro with kernel 6.1.23, X11 (*no wayland*) with libx11 1.8.4-1 as well, intel integrated, godot 4.1 compiled from source (from a fork not far from master, but judging from this report the issue is in godot, I could confirm if necessary), backtrace is exactly the same as @Eraph above.

Also, I noticed this is with .NET 7.0.3, while if I download the official stable godot mono from godotengine.org (not from Manjaro's pacman) it never crashes this way, and it's on .NET6, not sure if it matters.

Any other info I could provide to help debug this?

**Comment by ju5tevg3niy** [comment]
Same issue.
```
godot:  4.0.2.stable.official.7a0977ce2
os:     Gentoo
kernel: 6.1.22
de:     Xfce 4.18 / X11
render: Vulkan API 1.3.230 - Forward Mobile - Using Vulkan Device #0: Intel - Intel(R) HD Graphics 620 (KBL GT2)
```

**Comment by comminux** [comment]
It seems that the problem no longer occurs in version 1.8.5 (Arch Linux official extra repository).

**Comment by jivvy** [comment]
Just had this happen

swaywm
Arch Linux
AMD 6700XT
Godot v4.2.dev.custom_build [f8dbed4d0]
libx11 1.8.6-1

```
[xcb] Unknown request in queue while dequeuing
[xcb] You called XInitThreads, this is not your fault
[xcb] Aborting, sorry about that.
godot.linuxbsd.editor.x86_64: xcb_io.c:175: dequeue_pending_request: Assertion `!xcb_xlib_unknown_req_in_deq' failed.
```

**Comment by krendil** [comment]
I'm getting the same error, with a recent libX11 and non-Arch Linux

Godot Engine v4.1.1.stable.custom_build
OpenGL API 4.6 (Core Profile) Mesa 23.1.3 - Compatibility - Using Device: AMD - AMD Radeon RX 6600 (navi23, LLVM 15.0.7, DRM 3.52, 6.3.13_1)

Void Linux
XFCE4 / xfwm 4.18.0_1
libX11 1.8.6_1
libxcb 1.16_1

```
[xcb] Unknown request in queue while dequeuing
[xcb] You called XInitThreads, this is not your fault
[xcb] Aborting, sorry about that.
godot: xcb_io.c:175: dequeue_pending_request: Assertion `!xcb_xlib_unknown_req_in_deq' failed.
```

**Comment by Leshy-YA** [comment]
Confirmed on Fedora running KDE with Mesa Intel® Xe Graphics.
It would seem there's a regression in libX11 1.8.7, probably related to https://gitlab.freedesktop.org/xorg/lib/libx11/-/issues/170
Downgrading libX11 to 1.8.4 removes the issue.

**Comment by ZwieBit** [comment]
Can also confirm that a downgrade to libX11 1.8.4 fixed the issue. I already thought that godot is somewhat unstable but now even 4.2 beta1 works like a charm :-)

**Comment by vvvvvvitor** [comment]
Keeps happening here all the time, it makes editor basically unusable due to how often it happens. It's really frustrating to the point of me not wanting to work on my project anymore.

`[xcb] Unknown sequence number while awaiting reply
[xcb] You called XInitThreads, this is not your fault
[xcb] Aborting, sorry about that.
4.1.3.x86_64: xcb_io.c:374: poll_for_response: Assertion `!xcb_xlib_threads_sequence_lost' failed.
`

**Comment by ygingras** [comment]
On Ubuntu 23.04, all of Godot 4.1.1, 4.1.2, 4.1.3 crash about three times per hour. If I downgrade xserver-xorg-core to 2:21.1.7-1ubuntu3.1, then the crashes happen only once every few days.

**Comment by vvvvvvitor** [comment]
> On Ubuntu 23.04, all of Godot 4.1.1, 4.1.2, 4.1.3 crash about three times per hour. If I downgrade xserver-xorg-core to 2:21.1.7-1ubuntu3.1, then the crashes happen only once every few days.

Single window mode also helps alot.

**Comment by akien-mga** [comment]
Please report the issue to https://gitlab.freedesktop.org/xorg/lib/libx11, it should not regress and break existing applications.

Even if we could find a workaround in Godot to not trigger whatever makes libx11 fail, all existing Godot releases and published games would still be broken. So libx11 needs to stop breaking Godot every other *patch* release.

**Comment by Leshy-YA** [comment]
> Please report the issue to https://gitlab.freedesktop.org/xorg/lib/libx11, it should not regress and break existing applications.

Reported: https://gitlab.freedesktop.org/xorg/lib/libx11/-/issues/199 I hope we can get a fix before Fedora 40.

**Comment by ilmagico** [comment]
Forgive my ignorance here, but given `libx11`'s instability as of recently, would it make sense to vendor and statically link a properly patched version of `libx11` in godot? Basically, if upstream cannot fix this (and/or keep it from regressing), can Godot take matters into its own hands? It won't fix existing released games, but it might fix new ones, and provide an easy way for devs to re-release a game with the fix. Just my 2c, not sure if this makes sense.

**Comment by Lamby777** [comment]
Yep, now I have a reason not to feel bad about catching myself subconsciously spamming CTRL+S.

Milestone shows 4.3, would the fix last or do you guys think an update gonna break it again soon after? Might just downgrade libx11 if it gets really annoying but I'm kinda too busy to troubleshoot stuff rn if doing so happens to break any of my other packages. ;-;

**Comment by granitrocky** [comment]
@Lamby777 Upgrading libx11 works, too. I compiled and installed the latest libx11 from master on fedora 39 by doing

`./autogen.sh`
`./configure --prefix=/usr`
`make`
`sudo make install`

and then reboot and I haven't had a crash yet.

**Comment by QueenOfSquiggles** [comment]
Gonna share I'm experiencing this regularly on my Manjaro KDE machine. 
System details here:
```
Operating System: Manjaro Linux 
KDE Plasma Version: 5.27.10
KDE Frameworks Version: 5.113.0
Qt Version: 5.15.11
Kernel Version: 6.5.13-7-MANJARO (64-bit)
Graphics Platform: X11
Processors: 4 × AMD Athlon(tm) X4 880K Quad Core Processor
Memory: 15.6 GiB of RAM
Graphics Processor: NVIDIA GeForce GTX 1070/PCIe/SSE2
Manufacturer: Gigabyte Technology Co., Ltd.
```
Currently working with Godot 4.2-Stable running it from the command line so I get more interesting details.

Error from terminal:
```
[xcb] Unknown sequence number while processing queue
[xcb] You called XInitThreads, this is not your fault
[xcb] Aborting, sorry about that.
Godot_v4.2-stable_linux.x86_64: xcb_io.c:278: poll_for_event: Assertion `!xcb_xlib_threads_sequence_lost' failed.
zsh: IOT instruction (core dumped)  $GODOT4_BIN -e .
```

**Comment by acolbert1986** [comment]
System info...
```
Operating System: Manjaro Linux 
KDE Plasma Version: 5.27.10
KDE Frameworks Version: 5.113.0
Qt Version: 5.15.11
Kernel Version: 6.1.69-1-MANJARO (64-bit)
Graphics Platform: X11
Processors: 12 × 12th Gen Intel® Core™ i5-12500T
Memory: 7.5 GiB of RAM
Graphics Processor: Mesa Intel® UHD Graphics 770
Manufacturer: Dell Inc.
Product Name: OptiPlex 3000
```

Error in terminal...
```
Godot Engine v4.2.1.stable.arch_linux - https://godotengine.org

[xcb] Unknown request in queue while dequeuing
[xcb] You called XInitThreads, this is not your fault
[xcb] Aborting, sorry about that.
godot: xcb_io.c:175: dequeue_pending_request: Assertion `!xcb_xlib_unknown_req_in_deq' failed.
[1]    2223 IOT instruction (core dumped)  godot --editor --verbose --single-window
```

**Comment by akien-mga** [comment]
> @Lamby777 Upgrading libx11 works, too. I compiled and installed the latest libx11 from master on fedora 39 by doing
> 
> https://gitlab.freedesktop.org/xorg/lib/libx11/-/tree/master
> 
> `./autogen.sh` `./configure --prefix=/usr` `make` `sudo make install`
> 
> and then reboot and I haven't had a crash yet.

That's weird, as the latest `master` commit of libx11 *is* the 1.8.7 release which is shipped by Fedora.

So there's a difference between the Fedora package for libX11-1.8.7-1.fc39 and the one you compiled locally. It can be a different set of install dependencies so that your local builds adds or removes a feature, or this patch that Fedora is chugging along https://src.fedoraproject.org/rpms/libX11/blob/rawhide/f/dont-forward-keycode-0.patch, or any of the other custom tweaks in their .spec file, though I don't see much that sounds relevant: https://src.fedoraproject.org/rpms/libX11/blob/rawhide/f/libX11.spec

Either way, I suggest also opening a Fedora bug report, as the upstream libX11 report isn't getting any traction.

**Comment by Leshy-YA** [comment]
> Either way, I suggest also opening a Fedora bug report, as the upstream libX11 report isn't getting any traction and we now have evidence that a self-compiled libx11 performs differently.

Done: https://bugzilla.redhat.com/show_bug.cgi?id=2256495

**Comment by ionenwks** [comment]
Possible CFLAGS are involved and some UB is causing this? aka when you built libX11 manually from master I'd assume you didn't use the same (possibly just a plain -O2). I don't really keep up with fedora, but believe they use LTO nowadays for one?

Haven't run into crashes myself on Gentoo, albeit I may not use it enough to run into these (I just test godot a bit for packaging, haven't got bug reports either way).

**Comment by akien-mga** [comment]
@ionenwks That's a good call. Here are the build flags used on Fedora (as of Fedora 38 on my VM, but it's likely similar on F39).

```
$ rpm --eval %build_cflags
-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64  -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer 

$ rpm --eval %build_cxxflags
-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64  -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer 

$ rpm --eval %build_ldflags
-Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -Wl,--build-id=sha1
```

**Comment by granitrocky** [comment]
@ionenwks 

Interestingly, when I run that locally, I see that there are some flags with multiple spaces between them. Note `-m64`
in `build_cflags` having a lot of extra whitespace surrounding it. I assume that's fine for the compiler, but maybe it's not?


`$ rpm --eval %build_cflags 
-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64   -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer 

$ rpm --eval %build_cxxflags
-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64   -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer  

$ rpm --eval %build_ldflags
-Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -Wl,--build-id=sha1  
`

**Comment by akien-mga** [comment]
@granitrocky That's not a problem for the compiler, and that's expected. It's just an artifact of Fedora's RPM macro usage, some of which might be empty. Something like `%{?conditional_macro} %{?conditional_macro2} %{?conditional_macro3}` can definitely lead to two consecutive empty spaces if `conditional_macro2` is empty.

Anyway, that's just a nerdy RPM packager tangent, so I'll mark those two comments as off-topic ;)

**Comment by ZwieBit** [comment]
Just another confirmation: After having crashes every 10 minutes, i'm using godot editor now for almost 4 hours without any issues after compiling libx11 by myself.

**Comment by novalis** [comment]
(FWIW, this isn't RedHat-specific: I just got it with Debian, using libx11 1.8.7-1.  FWIW I'm running Wayland.)

The code in `xcb_io.c` looks like some hairy thing that's trying hard to be thread-safe. But if it's not successful, then this is the sort of error one might expect to see. And certainly different compilation options could affect how aggressively stuff gets reordered, which could affect thread safety.  So I thought, why doesn't Godot just put a mutex around the call?  And then I looked, and it already had -- in most cases.

But consider the following rule (from `xcb_io.c`):
>  * A single thread cannot be both the the event-reading and the
>  * reply-reading thread at the same time.

So we would expect the call to `XQueryTree` (which, inside libX11, calls `_XReply`, which seems to be "reply-reading") to also lock the same mutex -- but [it doesn't](https://github.com/godotengine/godot/blob/bb6b06c81343073f10cbbd2af515cf0dac1e6549/platform/linuxbsd/x11/display_server_x11.cpp#L6233). 

So it seems possible that there's a race there.  `XGetWindowProperty` is another possible culprit (the ones in `screen_get_usable_rect` only) . And XGetInputFocus. I haven't fully audited to see if there are other cases than these three.  

I have only read the code, so this could be totally bogus.  But it seems plausible.

**Comment by Lamby777** [comment]
> @Lamby777 Upgrading libx11 works, too. I compiled and installed the latest libx11 from master on fedora 39 by doing
> 
> https://gitlab.freedesktop.org/xorg/lib/libx11/-/tree/master
> 
> `./autogen.sh` `./configure --prefix=/usr` `make` `sudo make install`
> 
> and then reboot and I haven't had a crash yet.

it's been annoying me so much that i finally decided to go looking for this thread again... :P

sadly, it doesn't work :(
![](https://media1.tenor.com/m/itt8u63wFdAAAAAd/open-source-cage-the-elephant.gif)

Not that compiling your own version doesn't work; that I don't know. The actual compiling part doesn't work. I tried `./autogen.sh` and it was giving some error about xorg macros not being installed so i installed this package called `xorg-util-macros` and now it's complaining about some macro `XTRANS_CONNECTION_FLAGS` being possibly undefined... Is the macro package I installed just outdated? Cuz i just pulled libx11 source from master so maybe they changed some macros that haven't been put onto arch repos yet. Is that even the right package to install? Seems to be, since the error's gone, but idk

**Comment by Lamby777** [comment]
At least the error message apologizes, which I found somewhat amusing.

**Comment by imaducklol** [comment]
I'll add another data point I guess, got the same crash on two different machines:
Godot 4.2.1, Fedora 39 (Gnome 45 on Wayland), libX11 1.8.7,  i3-12100F, 5700xt, (Running godot under xwayland)
Godot 4.2.1, Arch (Gnome 46 on Xorg), libX11 1.8.9, i7-1185G7, Iris Xe, (Running godot under xorg)

Happened on both machines at around twice per hour. Wasn't able to pick up on anything specifically that caused them. 
I'll report back if I compile libX11 from source and that fixes anything.

**Comment by Tichau** [comment]
Don't know if it's useful, but here is a new repro:

> [xcb] Unknown request in queue while dequeuing
> [xcb] You called XInitThreads, this is not your fault
> [xcb] Aborting, sorry about that.
> TheGuild: ../../src/xcb_io.c:175: dequeue_pending_request: Assertion `!xcb_xlib_unknown_req_in_deq' failed.
> [1]    631819 segmentation fault  ./Path/To/Exe

Debian 12
Gnome 43.9
X.Org version: 1.22.1.9

If any other info is needed, I can edit this post.

**Comment by akien-mga** [comment]
> https://github.com/godotengine/godot/blob/be56cab58c056c074d1e02cd0b38641204e39f41/platform/linuxbsd/x11/dynwrappers/xlib-so_wrap.c#L6
> 
> That version is almost five years old. Is it possible that we're looking at some ABI incompatibility here?

That's a good question.

This hypothesis could be tested by someone who can reproduce the issue reliably, by making a custom build with `scons use_sowrap=no`, which will disable the dynamic library wrappers and link the system libraries instead. To compile successfully, you might need to install more dev libraries (the ones from https://docs.godotengine.org/en/latest/contributing/development/compiling/compiling_for_linuxbsd.html#distro-specific-one-liners, which wasn't updated now that we default to dlopen'ing these deps).

**Comment by ttencate** [comment]
Summarizing the reports above:

* libx11 1.8.2 - OK (unless patched)
* libx11 1.8.3 - broken
* libx11 1.8.4 - broken
* libx11 1.8.5 - OK (but see below)
* libx11 1.8.6 - broken
* libx11 1.8.7 - broken
* libx11 1.8.8 - unknown
* libx11 1.8.9 - broken

The only difference between 1.8.5 and 1.8.6 is [304a654](https://gitlab.freedesktop.org/xorg/lib/libx11/-/commit/304a654a0d57bf0f00d8998185f0360332cfa36c), which seems unrelated to me. So I'm inclined to assume that there was only one breakage, not two, and 1.8.5 is broken as well.

I tried rebuilding the Arch package from the [official PKGBUILD](https://gitlab.archlinux.org/archlinux/packaging/packages/libx11/-/blob/main/PKGBUILD). Even with this, I could not trigger the crash! **So for Arch users, this is a local workaround.** After reinstalling the official binary package, I got a crash within a minute or two.

The two `libX11.so.6.4.0` files are indeed different, but I can't tell if the differences are meaningful. Addresses and orders are different, but the list of exported symbols is the same. The two `libX11-xcb.so.1.0.0` files are the same size (13976 bytes), and the diff is small:
```
--- official-xcb.hex	2024-05-29 12:23:39.711349095 +0200
+++ mine-xcb.hex	2024-05-29 12:23:45.944791641 +0200
@@ -45,8 +45,8 @@
 000002c0: 0300 0000 0000 0000 0100 01c0 0400 0000  ................
 000002d0: 0100 0000 0000 0000 0200 01c0 0400 0000  ................
 000002e0: 0000 0000 0000 0000 0400 0000 1400 0000  ................
-000002f0: 0300 0000 474e 5500 cf41 1b11 8b82 2d5d  ....GNU..A....-]
-00000300: ad43 7b2d 9bad ab79 884a bc70 0000 0000  .C{-...y.J.p....
+000002f0: 0300 0000 474e 5500 7e7a c198 eaf0 26ab  ....GNU.~z....&.
+00000300: 1636 87ec 7be5 6f04 a5b9 943b 0000 0000  .6..{.o....;....
 00000310: 0200 0000 0500 0000 0100 0000 0600 0000  ................
 00000320: 0000 0200 0005 0008 0500 0000 0600 0000  ................
 00000330: 6be4 cc2e 3b9a cb9a 0000 0000 0000 0000  k...;...........
@@ -767,10 +767,10 @@
 00002fe0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
 00002ff0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
 00003000: 0040 0000 0000 0000 4743 433a 2028 474e  .@......GCC: (GN
-00003010: 5529 2031 332e 322e 3120 3230 3233 3038  U) 13.2.1 202308
-00003020: 3031 0000 6c69 6258 3131 2d78 6362 2e73  01..libX11-xcb.s
+00003010: 5529 2031 342e 312e 3120 3230 3234 3035  U) 14.1.1 202405
+00003020: 3037 0000 6c69 6258 3131 2d78 6362 2e73  07..libX11-xcb.s
 00003030: 6f2e 312e 302e 302e 6465 6275 6700 0000  o.1.0.0.debug...
-00003040: 164f e2c3 002e 7368 7374 7274 6162 002e  .O....shstrtab..
+00003040: fdef bdc9 002e 7368 7374 7274 6162 002e  ......shstrtab..
 00003050: 6e6f 7465 2e67 6e75 2e70 726f 7065 7274  note.gnu.propert
 00003060: 7900 2e6e 6f74 652e 676e 752e 6275 696c  y..note.gnu.buil
 00003070: 642d 6964 002e 676e 752e 6861 7368 002e  d-id..gnu.hash..
```

This does give a clue: apparently the official binary package was compiled with GCC 13.2.1, whereas I'm using GCC 14.1.1. This explains the differences in `libX11.so.6.4.0` as well. But I don't think GCC is to blame here – it's probably just a subtle difference that causes the actual (probably thread-related) bug to manifest or not.

Not being able to reproduce this in my own build, even before adding debug information, makes this thing very hard to debug, but I'll keep trying.

**Comment by ttencate** [comment]
I installed the `gcc13` package and used it to compile libX11 again from the official PKGBUILD, but modified with `CC=gcc-13 CPP=cpp-13 AR=gcc-ar-13 NM=gcc-nm-13 RANLIB=gcc-ranlib-13` before the `./configure` command. (Not sure all of these are necessary or even correct; `CC` is the main one.) Even this didn't help to reproduce the crash.

---

Something I found in the core dump: at the time of the crash, there were _two_ threads interacting with xcb. The main thread, that aborted:

```
...
#19 0x00007418b78c1c67 in __assert_fail (
    assertion=assertion@entry=0x7418b6d64528 "!xcb_xlib_unknown_req_in_deq", 
    file=file@entry=0x7418b6d644df "xcb_io.c", line=line@entry=175, 
    function=function@entry=0x7418b6d77310 <__PRETTY_FUNCTION__.6> "dequeue_pending_request") at assert.c:103
#20 0x00007418b6cfbcef in dequeue_pending_request (dpy=dpy@entry=0x62c40271a710, req=req@entry=0x74187000c270)
    at /usr/src/debug/libx11/libX11-1.8.9/src/xcb_io.c:175
#21 0x00007418b6cfec95 in poll_for_response (dpy=dpy@entry=0x62c40271a710)
    at /usr/src/debug/libx11/libX11-1.8.9/src/xcb_io.c:381
#22 0x00007418b6d019b2 in _XEventsQueued (dpy=0x62c40271a710, mode=<optimized out>)
    at /usr/src/debug/libx11/libX11-1.8.9/src/xcb_io.c:441
#23 0x00007418b6cdecdf in XFlush (dpy=0x62c40271a710) at /usr/src/debug/libx11/libX11-1.8.9/src/Flush.c:39
#24 0x000062c3f636ae5c in DisplayServerX11::_wait_for_events (this=this@entry=0x62c4026fbe50)
    at platform/linuxbsd/x11/display_server_x11.cpp:4048
#25 0x000062c3f636d070 in DisplayServerX11::_poll_events (this=0x62c4026fbe50)
    at platform/linuxbsd/x11/display_server_x11.cpp:4074
#26 0x000062c3f9fc3e2d in Thread::callback (p_caller_id=<optimized out>, p_settings=..., 
    p_callback=0x62c3f636d0b0 <DisplayServerX11::_poll_events_thread(void*)>, p_userdata=0x62c4026fbe50)
    at core/os/thread.cpp:61
#27 0x000062c3fa8a60e4 in execute_native_thread_routine ()
#28 0x00007418b791fded in start_thread (arg=<optimized out>) at pthread_create.c:447
#29 0x00007418b79a30dc in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:78
```

And a thread that appears to belong to AMD's Vulkan driver:

```
#0  0x00007418b799539d in __GI___poll (fds=fds@entry=0x74187edffae8, nfds=nfds@entry=1, 
    timeout=timeout@entry=-1) at ../sysdeps/unix/sysv/linux/poll.c:29
#1  0x00007418b6ca420b in poll (__timeout=-1, __nfds=1, __fds=0x74187edffae8) at /usr/include/bits/poll2.h:39
#2  _xcb_conn_wait (c=c@entry=0x62c40271b9d0, vector=vector@entry=0x0, count=count@entry=0x0, 
    cond=<optimized out>) at /usr/src/debug/libxcb/libxcb-1.17.0/src/xcb_conn.c:510
#3  0x00007418b6ca629b in _xcb_conn_wait (count=0x0, vector=0x0, cond=<optimized out>, c=0x62c40271b9d0)
    at /usr/src/debug/libxcb/libxcb-1.17.0/src/xcb_conn.c:476
#4  xcb_wait_for_special_event (c=0x62c40271b9d0, se=0x62c402b86190)
    at /usr/src/debug/libxcb/libxcb-1.17.0/src/xcb_in.c:806
#5  0x00007418a31c18f0 in ?? () from /usr/lib/amdvlk64.so
#6  0x00007418a31bd495 in ?? () from /usr/lib/amdvlk64.so
#7  0x00007418a31df714 in ?? () from /usr/lib/amdvlk64.so
#8  0x00007418a322ed61 in ?? () from /usr/lib/amdvlk64.so
#9  0x00007418b791fded in start_thread (arg=<optimized out>) at pthread_create.c:447
#10 0x00007418b79a30dc in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:78
```

The latter is hanging in a `poll` call, so it wasn't actively racing at the time of the crash, but it's an interesting tidbit that might be a reason why Godot suffers from this bug and other applications don't. I tried with the `vkcube` spinning cube Vulkan demo; I couldn't get it to crash, but upon killing it with `SIGQUIT` (Ctrl+\), this shows the same amdvlk backtrace in its coredump as well.

**Comment by ttencate** [comment]
Line numbers refer to libx11 1.8.9, although the file `src/xcb_io.c` hasn't been touched in two years.

[On line 319](https://gitlab.freedesktop.org/xorg/lib/libx11/-/blob/97fb5bda3d0777380cd4b964f48771a82ef3f2a7/src/xcb_io.c#L319) in `poll_for_response()`, we set:
```
        req = dpy->xcb->pending_requests;
```

There is no code that modifies the `req` pointer in the meantime. Then, if there is actually a pending request and some other conditions hold, the pending requests is dequeued:

```
        dequeue_pending_request(dpy, req);
```

And the first thing that function does, is to fail the assertion:
```
    if (req != dpy->xcb->pending_requests)
        throw_thread_fail_assert("Unknown request in queue while "
                                 "dequeuing",
                                 xcb_xlib_unknown_req_in_deq);
```

Since `req` is a local variable and hasn't been changed, this must mean that `dpy->xcb->pending_requests` has been changed in the meantime. The culprit must have been either some invalid memory access on the same thread, or a race condition from a different thread. My money is on the latter. (It could theoretically also have been some callback that performed a reentrant libx11 call, but I don't see any place where callbacks are invoked here; also, it would imply a lack of locking somewhere, same as a threading issue.)

It should be noted that we are in an `XFlush()` call, which is a critical section, calling `LockDisplay()` at the start and `UnlockDisplay()` at the end. So if this is a threading issue, we'd want to look for places that modify `pending_requests` without issuing such a lock.

There are only two such places that matter: `append_pending_request` and `dequeue_pending_request`. So I set a conditional breakpoint in both, with the condition `dpy->lock->mutex->__data->__owner == 0` (relying on some pthread internals to check if the mutex is locked). After a few minutes, the breakpoint was hit, yielding the following stack trace:

```
#0  dequeue_pending_request (dpy=dpy@entry=0x55555cd1fde0, req=req@entry=0x55556a1df6f0)
    at /usr/src/debug/libx11/libX11-1.8.9/src/xcb_io.c:174
#1  0x00007ffff7103343 in _XReply (dpy=0x55555cd1fde0, rep=0x7fffffffdb00, extra=0, discard=0)
    at /usr/src/debug/libx11/libX11-1.8.9/src/xcb_io.c:736
#2  0x00007ffff70e40f4 in XGetWindowProperty (dpy=0x55555cd1fde0, w=25165826, property=372, offset=0, 
    length=32, delete=<optimized out>, req_type=4, actual_type=0x7fffffffdbb8, actual_format=0x7fffffffdbb4, 
    nitems=0x7fffffffdbc0, bytesafter=0x7fffffffdbc8, prop=0x7fffffffdbd0)
    at /usr/src/debug/libx11/libX11-1.8.9/src/GetProp.c:69
#3  0x0000555555af1360 in DisplayServerX11::_window_minimize_check (this=this@entry=0x55555ccfc9f0, 
    p_window=p_window@entry=0) at platform/linuxbsd/x11/display_server_x11.cpp:2375
#4  0x0000555555af167f in DisplayServerX11::window_get_mode (this=0x55555ccfc9f0, p_window=0)
    at platform/linuxbsd/x11/display_server_x11.cpp:2705
#5  0x0000555555aeba48 in DisplayServerX11::can_any_window_draw (this=0x55555ccfc9f0)
    at platform/linuxbsd/x11/display_server_x11.cpp:2912
#6  0x0000555555b45426 in Main::iteration () at main/main.cpp:3685
#7  0x0000555555ad7311 in OS_LinuxBSD::run (this=this@entry=0x7fffffffddb0)
    at platform/linuxbsd/os_linuxbsd.cpp:958
#8  0x0000555555ac5176 in main (argc=<optimized out>, argv=0x7fffffffe398)
    at platform/linuxbsd/godot_linuxbsd.cpp:74
```

When continuing the program after the breakpoint is hit, it immediately crashes apologetically.

The API function `XGetWindowProperty` called from Godot does lock the mutex, but `_XReply` [transiently unlocks it for a while](https://gitlab.freedesktop.org/xorg/lib/libx11/-/blob/97fb5bda3d0777380cd4b964f48771a82ef3f2a7/src/xcb_io.c#L701-706). And apparently, by the time `dequeue_pending_request` is called [here](https://gitlab.freedesktop.org/xorg/lib/libx11/-/blob/97fb5bda3d0777380cd4b964f48771a82ef3f2a7/src/xcb_io.c#L736), the mutex is somehow not locked.

This is as far as I got for today. I tried setting more breakpoints in `_XReply` to find out where the lock is lost, but the breakpoints all end up a the top of the function for some reason, and might also interfere with my ability to trigger the crash. Stupid Heisenbug.

**Comment by Some1and2-XC** [comment]
Hey, this seems to still be an issue running on Ubuntu 24.04 running x11 and KDE.
```sh
Godot Engine v4.3.stable.official.77dcf97d8 - https://godotengine.org
Vulkan 1.3.274 - Forward+ - Using Device #0: Intel - Intel(R) UHD Graphics (ICL GT1)

[xcb] Unknown request in queue while dequeuing
[xcb] You called XInitThreads, this is not your fault
[xcb] Aborting, sorry about that.
project.x86_64: ../../src/xcb_io.c:175: dequeue_pending_request: Assertion `!xcb_xlib_unknown_req_in_deq' failed.
Aborted (core dumped)
```

I wonder if this is an intel integrated graphics thing? Also should I try and update something to fix this?

**Comment by novalis** [comment]
It's not specific to Intel -- I have an AMD Radeon and I get it.  

Nobody seems to have found a fix yet, so don't bother updating anything.

**Comment by Invium-GH** [comment]
> Summarizing the reports above:
> 
> * libx11 1.8.2 - OK (unless patched)
> * libx11 1.8.3 - broken
> * libx11 1.8.4 - mixed
> * libx11 1.8.5 - OK (but see below)
> * libx11 1.8.6 - broken
> * libx11 1.8.7 - broken
> * libx11 1.8.8 - unknown
> * libx11 1.8.9 - broken
>
1.8.10 - Broken.

I have been struggling with this very same crash for some time now in my personal project and have also resorted to trying to create an MRP to force the crash. The intent being that if I could find a force for the crash, Godot's developers could maybe use that information to find a workaround.

So far, no luck in forcing it. Will continue to attempt it though.

Also, so far I have been unable to eradicate it completely from my project with workarounds. Although I do think I have managed to reduce the occurrence rate. But this could just be a feeling. As the crash is very random in nature. Very hard to predict and to reproduce.

**Comment by Invium-GH** [comment]
One thing I have noticed while struggling with finding a workaround or MRP to force the crash is that I have yet to see it happen with the editor or the running project being in focus.

I have seen both the running project as well as the editor itself (without the project running) crash as a result of this issue but never when in focus. Always when focus was on a different task. Most recent crash I've had was with Godot being open, my project not running and me just working on some C# code for the project, using an external editor (VSCode).

And any time my project itself runs into this crash it is with the project not being in focus.

Maybe this could help to guide someone more familiar with the interaction between Godot and the crashing library towards a workaround.

For example, is there something different between how Godot updates/draws windows which are in focus compared to how it updates/draws windows which are not in focus?

**Comment by Calinou** [comment]
> For example, is there something different between how Godot updates/draws windows which are in focus compared to how it updates/draws windows which are not in focus?

The editor is capped to 10 FPS by default when unfocused. You can disable this behavior by enabling **Update Continuously** in the Editor Settings.

**Comment by Invium-GH** [comment]
> You can disable this behavior by enabling **Update Continuously** in the Editor Settings.

I haven't seen the editor crash since enabling this. It's been a total uptime of at least a few dozen hours by now. So, it would appear there is _something_ to the crash being more likely to happen with infrequently updated windows or windows with a lower frame rate. But, this is hardly a viable workaround. As per the tooltip of this option, it is more to be used for troubleshooting purpses.

Incidentally, is there something else about _projects_ run in the background versus in focus? Like, some internal optimization the engine performs to save on CPU/GPU cycles for unfocused windows? So, not related to the editor but related to the engine itself.

**Comment by Invium-GH** [comment]
Seeing as how the issue is a bug in xcb or seems to be anyhow, there is nothing really profound we can do in Godot to mitigate it. I've been thinking about this and all I can think of to do is to simply always assume _false_ when checking for LowCPUMode when running on X11. In both the editor as well as the engine itself. Had a look at the code myself but, well, it's a big project and to find the most clean, optimal and efficient way to do this might require more intimate knowledge of the engine and the editor's source code.

I think it should go in core/os/os.cpp, around line 146:

```bool OS::is_in_low_processor_usage_mode() const {
	return low_processor_usage_mode;
}```

As in, turning that into a ternary. I'm just not sure... what to use there.

So...

```bool OS::is_in_low_processor_usage_mode() const {
	return **_X11 ? false :_** low_processor_usage_mode;
}```

**Comment by novalis** [comment]
Please see my comment in <https://github.com/godotengine/godot/issues/75308#issuecomment-1965725917> -- there is a reason to believe this is a mistake in Godot, and no reason to believe it's an xcb bug. Having read the xcb code, it's pretty straightforward and unlikely to be the source of the bug.

Before messing with random Godot things that are merely correlations, it would make sense to add a few more mutexes and see if that helps.

**Comment by Invium-GH** [comment]
> it would make sense to add a few more mutexes and see if that helps.

Which is what I tried to do with my inquiry here: https://github.com/godotengine/godot/issues/75308#issuecomment-2626197891

Which went entirely unanswered. Leading me to simply look for possible workarounds. I'd rather see a temporary quirk being added to the engine and the problem being _avoided_ that way than people looking for a year or more in search of the actual fix. But maybe that's just me.

Thing is, just to test, I turned _Update continuously_ off and within mere hours, it crashed again. I enable it again, no more crashes. The workaround _would_ at least work to prevent crashes on X11. Whether it's as a result of xcb or Godot, a temporary quirk like that would allow both the engine as well as the editor to run reliably.

Do I pretend to understand the crash? No. But I cannot ignore the correlation between LowCPUMode and the crash.

**Comment by akien-mga** [comment]
The fact that low CPU usage mode seems to trigger it for the editor is an interesting find, but indeed disabling support for it on X11 is not a solution. This would break a number of non-game uses (not just the Godot editor but other apps using it), and would lead to excess CPU usage / battery drain on laptop.

It's a good workaround to know of for users who are affected though (which aren't all Linux users - I've never managed to reproduce that crash myself on any of my Linux computers running KDE, neither on native X11 nor Xwayland).

**Comment by hhyyrylainen** [comment]
I've seen this crash a couple of times on my computer after updating to Fedora 41. Before that I don't really recall seeing this problem, so it could be some update in Fedora 41 that makes this issue more likely to surface. The game I'm working on doesn't have low CPU mode set. 

For me I've seen this crash much more rarely affect the Godot Editor and mostly the game when running sometimes crashing like this. We've also had reports from players with the same crash (a few over the past 4-5 months) who are playing the exported version of the game. Most recent report we got about someone seeing this crash is here: https://github.com/Revolutionary-Games/Thrive/issues/5937

**Comment by mouseroot** [comment]
Lubuntu: `uname -a` 
```Linux Alpha 6.11.0-18-generic #18-Ubuntu SMP PREEMPT_DYNAMIC Fri Feb  7 22:34:25 UTC 2025 x86_64 x86_64 x86_64 GNU/Linux```

Also get random crashes with the error
```
[xcb] Unknown sequence number while processing queue
[xcb] You called XInitThreads, this is not your fault
[xcb] Aborting, sorry about that.
Godot 4.3.bin: ../../src/xcb_io.c:278: poll_for_event: Assertion `!xcb_xlib_threads_sequence_lost' failed.
```

This is also in Godot 4.3

**Comment by AimForNaN** [comment]
How many linux gamers actually still use X11? Wouldn't the simplest solution be to make X11 opt-in? Wayland is currently opt-in. Just make Wayland the default and call it a day. This isn't technically an issue with Godot itself but simply making use of buggy and outdated dependencies that are barely being worked on anymore. A simple google search would reveal that much.

**Comment by Invium-GH** [comment]
> How many linux gamers actually still use X11?

Most of them. Since Wayland and WINE/Proton still do not really work as well together as X11 and WINE/Proton do.

> This isn't technically an issue with Godot itself

Actually, it is. I haven't seen any report of this crash occurring with any other software other than Godot, at least not for a few years by now. Yes, technically, it _probably_ is a problem in X11 _somewhere._ But... Godot is currently the only thing triggering it and the workaround for that is entirely based within Godot as well... disabling LowCPU mode.

Anyhow, forcing Wayland onto people that for whatever reason _do_ want to use Linux _and_ Godot is no solution either. Personally, I'd be forced to switch a different engine if that were ever to be the decision here. Wayland, on paper, is good. But it simply is not as mature as X11 yet. It needs some more time in the oven, for me personally. The experience may differ for others but, well, not everyone is as enthusiastic about Wayland as a Google search might imply.

Especially on Nvidia hardware the Wayland experience is far from ideal, primarily due to issues with the Nvidia GPU driver. And getting Nvidia to solve anything in a timely fashion, well... let's just say, we've seen how that ends. And would merely be shifting the responsibility.

Not a viable solution to ditch X11 completely. Not yet. Maybe in a year or 3.

**Comment by AimForNaN** [comment]
@Invium-GH 

> Most of them. Since Wayland and WINE/Proton still do not really work as well together as X11 and WINE/Proton do.

Just because most games are ran through XWayland does not mean that most people run on X11. Most popular distros use Wayland by default, and the average gamer uses popular distros. Proton works fine on Wayland.
 
> Actually, it is. I haven't seen any report of this crash occurring with any other software other than Godot, at least not for a few years by now. Yes, technically, it _probably_ is a problem in X11 _somewhere._ But... Godot is currently the only thing triggering it and the workaround for that is entirely based within Godot as well... disabling LowCPU mode.

A google search suggests that it is internal to X11, since it's not unique to Godot. I even know of cases regarding GTK. Sure, bandages can be applied, but why bother insisting on supporting legacy software?

> Anyhow, forcing Wayland onto people that for whatever reason _do_ want to use Linux _and_ Godot is no solution either. Personally, I'd be forced to switch a different engine if that were ever to be the decision here. Wayland, on paper, is good. But it simply is not as mature as X11 yet. It needs some more time in the oven, for me personally. The experience may differ for others but, well, not everyone is as enthusiastic about Wayland as a Google search might imply.

It might not be a solution for you, but it would definitely be a solution for Godot. If you stubbornly insist on X11, then opt-in and stick to older versions of Godot, and learn to save often. There's no point in supporting something that everyone is moving away from.

> Especially on Nvidia hardware the Wayland experience is far from ideal, primarily due to issues with the Nvidia GPU driver. And getting Nvidia to solve anything in a timely fashion, well... let's just say, we've seen how that ends. And would merely be shifting the responsibility.

IIRC, Nvidia jumped on the Wayland train since v555. The only reason this would be an issue is if you're on a distro that isn't a rolling release or if you rarely perform system updates. But that wouldn't be a Wayland or Nvidia problem.

> Not a viable solution to ditch X11 completely. Not yet. Maybe in a year or 3.

That's not what everyone else seems to think. And by everyone else I mean, popular distros, popluar desktop environments, popular GUI toolkits, among many other things. If it wasn't good enough, no one would be jumping on this train.

**Comment by gelvinp** [comment]
> The only reason this would be an issue is if you're on a distro that isn't a rolling release or if you rarely perform system updates.

Ok, hi. I am a real actual person who runs X11, specifically because Wayland is incredibly unresponsive and buggy for me, on a regularly updated arch Linux system. 

I give Wayland a try every few months, and I always switch back to X11 because it just plain works better on my computer. I’m glad that Wayland works for you, and when it eventually works as well for me as X11 does then I am willing to switch. That hasn’t happened yet however. I’m not being stubborn, Wayland _actually_ runs worse for me, and telling me to “learn to save often” is unhelpful and honestly pretty rude. 

Godot is the only piece of software that I have seen crash in this way, and I only started seeing it with version 4.X. Even if this is caused by some underlying x11 bug, something inside Godot changed to start hitting it, and I think it’s worthwhile to figure out what changed and find a solution, because I promise you that I am not the only person still using X11 because it’s genuinely a better experience for me.

**Comment by ttencate** [comment]
Whether and why people use X11 (I do too!) is not the point of the discussion here. If you use Wayland and all your users do too, great, you have no problem. But meanwhile, Godot supports X and has (as far as I know) no plans to drop support for it. So let's please remain ontopic and see if we can get this bug fixed, whether it's in Godot or in libX11 or in xcb or some combination thereof.

**And that starts with a reliable way to reproduce it.**

I myself (on Arch Linux) haven't seen this happen for a while. This _might_ correlate with the upgrade from libX11 1.8.10 to 1.8.11, but I don't see anything in its [commit history](https://gitlab.freedesktop.org/xorg/lib/libx11/-/commits/libX11-1.8.11?ref_type=tags) that jumps out to me as relevant. So I'm guessing it isn't fixed, and this is another case of "it was recompiled so now the UB manifests differently or not at all".

@AimForNaN, you mentioned seeing this in GTK apps as well. Do you have any references? Thus far, I had the impression that this was specific to Godot.

Meanwhile, I am running an empty Godot game 30 times in the background, capped at 5 fps and with low CPU mode enabled. All of them have been running without crashing all day...

**Comment by ju5tevg3niy** [comment]
It doesn't really matter which windowing system is used. The sporadic nature of this issue suggests that this is a datarace bug.
It might be within x11 client libraries, but it could also indicate an issue with godot's multithreading, some datarace/UB could just corrupt random data within process's memory but crash only during x11 code execution because it has asserts that maintain internal invariants.
Wayland libraries could also have bugs that are yet to be discovered. I haven't seen such crashes with other multithreaded projects utilizing x11 (e.g. firefox, blender, etc.).

> [@Invium-GH](https://github.com/Invium-GH)
> 
> > Most of them. Since Wayland and WINE/Proton still do not really work as well together as X11 and WINE/Proton do.
> 
> Just because most games are ran through XWayland does not mean that most people run on X11. Most popular distros use Wayland by default, and the average gamer uses popular distros. Proton works fine on Wayland.

It depends on which DE you use, not all of them support Wayland yet (e.g. xfce), not all distros have deployed the most latest versions of ones which support (e.g gnome, kde).

> > Actually, it is. I haven't seen any report of this crash occurring with any other software other than Godot, at least not for a few years by now. Yes, technically, it _probably_ is a problem in X11 _somewhere._ But... Godot is currently the only thing triggering it and the workaround for that is entirely based within Godot as well... disabling LowCPU mode.
> 
> A google search suggests that it is internal to X11, since it's not unique to Godot. I even know of cases regarding GTK. Sure, bandages can be applied, but why bother insisting on supporting legacy software?

X11 is far from being legacy software. Supporting it means godot could run on ~99% desktop linux systems. If the issue is actually internal to x11 then it must be fixed and it will benefit all of the ecosystem.

> > Anyhow, forcing Wayland onto people that for whatever reason _do_ want to use Linux _and_ Godot is no solution either. Personally, I'd be forced to switch a different engine if that were ever to be the decision here. Wayland, on paper, is good. But it simply is not as mature as X11 yet. It needs some more time in the oven, for me personally. The experience may differ for others but, well, not everyone is as enthusiastic about Wayland as a Google search might imply.
> 
> It might not be a solution for you, but it would definitely be a solution for Godot. If you stubbornly insist on X11, then opt-in and stick to older versions of Godot, and learn to save often. There's no point in supporting something that everyone is moving away from.

Asking others to use older versions is not a solution.

> > Especially on Nvidia hardware the Wayland experience is far from ideal, primarily due to issues with the Nvidia GPU driver. And getting Nvidia to solve anything in a timely fashion, well... let's just say, we've seen how that ends. And would merely be shifting the responsibility.
> 
> IIRC, Nvidia jumped on the Wayland train since v555. The only reason this would be an issue is if you're on a distro that isn't a rolling release or if you rarely perform system updates. But that wouldn't be a Wayland or Nvidia problem.

Not everyone uses rolling release distro or even the latest Fedora or Ubuntu. Stability matters when you're working on a long-term projects and don't want to bother with system tinkering or fixing something after a system update.

> > Not a viable solution to ditch X11 completely. Not yet. Maybe in a year or 3.
> 
> That's not what everyone else seems to think. And by everyone else I mean, popular distros, popluar desktop environments, popular GUI toolkits, among many other things. If it wasn't good enough, no one would be jumping on this train.

Distros, DEs and toolkits still support x11 and will support it for a long time because, again, it covers ~99% desktop linux systems.

**Comment by AimForNaN** [comment]
@gelvinp You say it's rude and unhelpful to require people to save often. Then don't save often. You be you. But you should find it equally rude to require developers to support legacy software for as long as you desire.

@ttencate Certain GTK apps running through XWayland crashing randomly went away after applying system updates a long time ago. Whether or not it was patched in GTK or if the app stopped running through XWayland, I don't know. While I do remember seeing bug reports for GTK, I didn't keep up with them.

@ju5tevg3niy It would matter if it turns out that it indeed has to do with the underlying window system and, therefore, would place it outside of the scope of Godot. Sure, it would benefit X11 if it is fixed, but if it's internal to X11, it wouldn't be Godot's burden. If it's, however, truly only Godot's fault, then debugging would be much simpler than trying to debug an entire underlying window system. It's been almost two years since this was last reported. While it is arguably a random occurrence, it appears to happen enough to be an issue. How is it that we have so many of you arguing in favor of X11, but there is still question as to where exactly is the problem? Maybe it's the case that most have opted in to Wayland.

Asking others to use older versions is a solution, it's just not one you agree with. Suggesting that X11 is stable is contradicted by this and other issues and by its deprecation. I never said distros don't support X11. The argument was which one they default to. Which would mean that their users would be on Wayland by default. And "popular" meaning that most Linux users are on those distros. It doesn't matter if not everyone uses them. The argument was if most gamers do. The only reason (popular) distros continue to support X11 isn't because they assume that most of their users are on X11. But GUI toolkits don't have to worry about supporting other applications, nor would any developer using their toolkit to build apps find much need to be mindful of X11. I don't buy this "99% of Linux users are X11 users" claim.

**Comment by akien-mga** [comment]
As @ttencate mentioned, discussion on X11 vs Wayland are off-topic and irrelevant here.

There's no intention to drop X11 support in the foreseeable future.

We may make the Wayland backend the default one (on Wayland only obviously) in the future when it has both feature parity in its Godot implementation, and good ecosystem support for drivers / real-time games. We're not there yet but again that's off-topic here.

Let's go back to the actual bug at hand. I'll hide the Wayland sidebar as off-topic.

**Comment by ttencate** [comment]
@gelvinp Whoooa... you're the MVP, that's _incredibly_ helpful! Have you already tried with that commit reverted on master (or 4.4-stable)?

@pouleyKetchoupp Any idea what in that commit could be causing this issue?

**Comment by ttencate** [comment]
Great work, @gelvinp! That definitely looks like the smoking gun. 

I went code spelunking in Xlib and **found what I believe to be the root cause**. I filed an issue with libx11: [Data race in X*IfEvent which might be causing "Assertion `!xcb_xlib_threads_sequence_lost' failed"](https://gitlab.freedesktop.org/xorg/lib/libx11/-/issues/233) Let's see if this gains any traction.

I'm not sure that this completely explains what we're seeing in Godot, because all calls to `X*IfEvent` are protected by our own `events_mutex`, which would also establish a happens-before relationship in the same way Xlib's own failing locking would. Maybe there are other Xlib calls that internally call one of the `X*IfEvent` functions?

Meanwhile, even _if_ a fix is released soon, it'll take years until it's deployed everywhere, so we want to look at a workaround in Godot as well. @gelvinp Have you also tried commenting out only the `_check_pending_events` call? That would amount to almost a revert of ba96fc6a4dff123c748d7c554147f8ade047f76c, but I think a bit of input delay is preferable over random crashes.

(Aside: Reading [The X New Developer’s Guide: Xlib and XCB](https://www.x.org/wiki/guide/xlib-and-xcb/) strongly suggests to me that using XCB directly is a much better fit for Godot than using Xlib (which nowadays is a wrapper around XCB). This would give us more precise control and the lowest possible latency. But it would be a serious project, and that's a can of worms better left closed for now.)

**Comment by gelvinp** [comment]
@ttencate After another 24 hour run, I can confirm that un-commenting the XFlush (so removing the `_check_pending_events` is the only change) does not cause it to start crashing again. Do you want me to open a PR with that now, or are we looking for a different workaround?

**Comment by ttencate** [comment]
As much as I'd like to see this resolved, I'm not in charge here :) Maybe @akien-mga can decide on this tradeoff?

**Comment by akien-mga** [comment]
@gelvinp Yes, please open a PR with your solution. It seems to be a workaround but it helps point at the source of the problem and we can evaluate whether to merge the workaround, or investigate further to find the root cause of the issue and how to poll events safely.

**Comment by anzhi0708** [comment]
Just encountered the same issue on Ubuntu 24.04 LTS. Lucky me — it looks like it's already been fixed.

**Comment by kq5fYWKaB9G1I5Pp5A1ta4zx9QenukU3** [comment]
Even though a (temporary) patch has been merged already, it is still worth it to investigate the issue further. To this end, I would like to share a workaround which, so far, has proven itself effective at eliminating the problem entirely.

However, and this is a major caveat... it is not viable in a gaming scenario. It would only be viable for applications (not games) created in Godot. In my humble opinion, it still does point at a possible genuine solution to the issue.

The solution:
```taskset -a -c0 <executable>```
As in... pinning all threads of an executable to the same core. Since using this, the application I created using Godot has not caused a single XinitThreads crash.

It should be obvious that it is not viable in a game setting, for performance reasons. However, I feel there could still be some genuine solution hidden in this discovery. Whether it is a basic race condition across threads which is causing this issue or something else, the fact I can circumvent it entirely by forcing everything onto a single core is indicative of something.

**Comment by ttencate** [comment]
For what it's worth, the latency reintroduced by the workaround of #106798 is actually noticeable. In my game, I rotate the camera by dragging the mouse. When dragging quickly, there are occasional noticeable hiccups that feel like frame drops – at irregular intervals, but about once every few seconds on average. (When dragging slowly, the effect is of course less noticeable.) However, the Godot profilers report that I'm still rendering at full fps.

This exists in 4.5.1 but not in 4.5.0-rc2, which is the latest I tested with. Narrowing it down specifically to this PR doesn't seem worth it, since we already know that it might reintroduce some input lag.

I also get the feeling that Godot's script editor sometimes feels more laggy when typing, but that could be a different issue.

Don't get me wrong: the lag is still preferable to random crashing! But indeed it's just a stopgap until a better fix can be found.
